feat(webhooks): link to event in slack webhook

mariusandra opened the pull request:
## Problem

There was no way to connect a Slack webhook with an actual event

## Changes

In Slack webhooks, in addition to `event.name`, `event.properties` and the undocumented `event.distinct_id`, this PR adds:

- `event.uuid`
- `event.event` (name + markdown link to `/events/:uuid`)

This also adds a redirect on `/events/:uuid` to `/events#q={query with hogql filter for event uuid}`.

## Caveat

The date filter is set to "last 24h". This works for recent Slack webhooks, but will be annoying for historic events. The proper fix is likely to include timestamps in the URL.

## How did you test this code?

- Visited `http://localhost:8000/events/01884f84-3c3d-0000-5881-66b1b7c64b20` and saw what I was looking for
- Added tests for the new webhook fields

mariusandra committed:
link to event

mariusandra committed:
fix test

mariusandra committed:
Merge branch 'master' into slack-event-link

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

mariusandra committed:
Merge branch 'master' into slack-event-link

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

mariusandra committed:
Merge branch 'master' into slack-event-link

thmsobrmlr commented on plugin-server/src/worker/ingestion/hooks.ts:
```diff
@@ -28,31 +28,36 @@ export function determineWebhookType(url: string): WebhookType {
+        markdown = `<${url}|${name}>`
     } else {
-        userMarkdown = `[${userName}](${siteUrl}/person/${event.distinctId})`
+        markdown = `[${name}](${url})`
```
This seems vulnerable to crafted events like `$pageview](https://malicious-url?`, which would then link to an arbitrary url in the Discord message.

mariusandra replied:
Hmm, true. This probably was an issue already before, but I added some escaping code now. Didn't test it yet 😅

mariusandra committed:
Merge branch 'master' into slack-event-link

mariusandra committed:
escape slack and markdown

mariusandra committed:
add some escaping

Twixes reviewed (changes requested):
Last 24h is very recent, I'm sure this will come up as an issue soon. Let's add the timestamp to the URL now.

mariusandra committed:
Merge branch 'master' into slack-event-link

mariusandra committed:
nicer escape

mariusandra committed:
timestamp in /events url

mariusandra committed:
add link to link to event

mariusandra commented:
I added some sanitisation to both the slack and markdown/teams webhooks. I hope it's at a "good enough" or "better than before" level, as it feels like a lot of work to test these locally. :)

I added timestamps into event URLs. To test them easily, I added a link to the event itself in the menu:
![2023-05-30 11 16 07](https://github.com/PostHog/posthog/assets/53387/297285ad-78f2-466a-9dce-16c325aeaa18)

thmsobrmlr reviewed (approved):
LGTM

Twixes commented on plugin-server/src/worker/ingestion/hooks.ts:
```diff
@@ -85,33 +129,34 @@ export function getValueOfToken(
-            text = stringify(event.event)
+        if (tokenParts[1] === 'uuid') {
+            markdown = text = webhookEscape(event.eventUuid, webhookType)
+        } else if (tokenParts[1] === 'name') {
+            markdown = text = webhookEscape(event.event, webhookType)
+        } else if (tokenParts[1] === 'event') {
+            ;[text, markdown] = getEventDetails(event, siteUrl, webhookType)
```
I think we should use the same approach as with `person` here. So `[event]` should be the linkified version, while `[event.event]` should return the name (and we probably should not use `[event.name]`, since in e.g. HogQL events have an `event` column and not `name`

mariusandra replied:
also makes sense. changed

mariusandra committed:
link to event instead

mariusandra committed:
use [event] instead

mariusandra commented on frontend/src/queries/nodes/DataTable/EventRowActions.tsx:
```diff
@@ -39,6 +39,17 @@ export function EventRowActions({ event }: EventActionProps): JSX.Element {
+                            fullWidth
+                            sideIcon={<IconLink />}
+                            data-attr="events-table-event-link"
+                            to={urls.event(event.uuid, event.timestamp)}
```
Good idea 👍 

<img width="309" alt="image" src="https://github.com/PostHog/posthog/assets/53387/04a4b237-432e-47bd-9c72-14a88d18d778">

Twixes reviewed (approved):
This is actually really useful, shipit!

mariusandra committed:
reasonably sure this will still be a string

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

mariusandra committed:
fix one more test

mariusandra committed:
Merge branch 'master' into slack-event-link

mariusandra committed:
escape
